src/3d/symbols/qgspointcloud3dsymbol_p.cpp

```diff
@@ -227,6 +267,179 @@ void QgsPointCloud3DSymbolHandler::makeEntity( Qt3DCore::QEntity *parent, const
+    const std::vector<size_t> &triangleIndexes = triangulation.triangles;
+    for ( size_t i = 0; i < triangleIndexes.size(); i += 3 )
+    {
+      QVector<QVector3D> triangleVertices( 3 );
+      for ( size_t j = 0; j < 3; ++j )
+      {
+        size_t vertIndex = triangleIndexes.at( i + j );
+        triangleVertices[j] = outNormal.positions.at( vertIndex );
+      }
+      //calculate normals
+      const QVector3D v1( triangleVertices.at( 1 ) - triangleVertices.at( 0 ) );
+      const  QVector3D v2( triangleVertices.at( 2 ) - triangleVertices.at( 0 ) );
+      const QVector3D partialNormal = QVector3D::crossProduct( v1, v2 );
+      for ( size_t j = 0; j < 3; ++j )
+        normals[triangleIndexes.at( i + j )] += partialNormal;
+    }
+
+    // Build now normals array
+    outNormal.normals.resize( ( outNormal.positions.count() ) * sizeof( float ) * 3 );
+    float *normPtr = reinterpret_cast<float *>( outNormal.normals.data() );
+    for ( int i = 0; i < normals.size(); ++i )
+    {
+      QVector3D normal = normals.at( i );
+      normal = normal.normalized();
+
+      *normPtr++ = normal.x();
+      *normPtr++ = normal.y();
+      *normPtr++ = normal.z();
+    }
+
+    // We apply a filter on triangles to improve the rendering:
+    // - we keep only triangles that have a least one point in the bounding box
+    // - if option is selected, we don't keep triangle with side greater than a threshold
+
+    outNormal.triangles.resize( triangleIndexes.size() * sizeof( quint32 ) );
+    quint32 *indexPtr = reinterpret_cast<quint32 *>( outNormal.triangles.data() );
+    size_t effective = 0;
+
+    bool sizeFilter = context.symbol()->filterTrianglesBySize();
+    bool heightFilter = context.symbol()->filterTrianglesByHeight();
+    float sizeThreshold =  context.symbol()->triangleSizeThreshold();
+    float heightThreshold =  context.symbol()->triangleHeightThreshold();
+
+    for ( size_t i = 0; i < triangleIndexes.size() / 3; ++i )
+    {
+      bool atLeastOneInBox = false;
+      bool greaterThanSize = false;
+      bool greaterThanHeight = false;
+      for ( size_t j = 0; j < 3; j++ )
+      {
+        QVector3D pos = outNormal.positions.at( triangleIndexes.at( i * 3 + j ) );
+        QVector3D pos2;
+
+        atLeastOneInBox |= bbox.intersects( pos.x(), pos.y(), pos.z() );
+
+        if ( heightFilter || sizeFilter )
+          pos2 = outNormal.positions.at( triangleIndexes.at( i * 3 + ( j + 1 ) % 3 ) );
+
+        if ( heightFilter )
+          greaterThanHeight |= std::fabs( pos.y() - pos2.y() ) > heightThreshold;
+
+        if ( sizeFilter && ! greaterThanHeight )
+        {
+          QVector3D pos2 = outNormal.positions.at( triangleIndexes.at( i * 3 + ( j + 1 ) % 3 ) );
+          // filter only in the horizontal plan, it is a 2.5D triangulation.
+          pos2.setY( 0 );
+          pos.setY( 0 );
+          greaterThanSize |= pos2.distanceToPoint( pos ) > sizeThreshold;
```

Comment:
I am wondering if it would easier to read if we simply used `(pos.x() - pos2.x())^2 + (pos.z() - pos2.z())^2 > sizeThresholdSquared` rather than setting y coords to zero...